base/nsjail-docker/Dockerfile

```diff
@@ -15,8 +15,9 @@ FROM gcr.io/kctf-docker/kctf-nsjail-chroot:latest AS chroot
 FROM gcr.io/kctf-docker/kctf-nsjail-bin:latest AS bin
 FROM ubuntu:19.10
 
-RUN apt-get update && apt-get install -y uidmap libprotobuf17 libnl-route-3-200
-RUN apt-get upgrade -y
```

Comment:
why no apt-get upgrade anymore? How do we get security fixes?

Reply:
We dont. Unfortunately `upgrade` itself anyway gets cached, which means that the image always has updates as of the **first** time the last person to push built the image.

There are some hacks to do cache busting (eg, adding an argument FOO and running `apt-get upgrade -y && echo $FOO` at the end of every Dockerfile), so that you always get the image as of the last time the last person to push built it. This has the slight disadvantage that the image will grow over time with things it wouldn't need if it used the "upstream" images, but worse of all, apparently upgrade can sometimes fail inside of docker (we need to build the containers with --privileged it seems).

But generally, the suggested approach to trully keep things up-to-date is a Rapid-style worker that rebuilds and pushes skipping cache (with `--pull --no-cache`)  daily or more often. For a CTF, I suspect this will have to be a one time operation before the competition that happens at release time? Although I don't know.

Note that GCR has a vulnerability scanner, that checks for outdated packages. Maybe we can just surface it's findings as warnings when users deploy outdated packages.

Reply:
Perhaps we can do a full build skipping cache when the challenges are marked as EXPOSED.

Maybe we can build the LoadBalancer (which anyway takes forever) earlier in the process, so that by the time we finish rebuilding, we hopefully don't have to wait so long.

Reply:
IMO upgrading once is still better than never at all.

We can have a LASTUPGRADE variable in the Dockerfile that is "never" by default.
Before the CTF starts, you just set this to a new value and check it into your git. All builds will then be at least that new.

Reply:
take a look at the docker documentation. it explains that apt-get upgrade can break when not running with --privileged (which we don't).

`Avoid RUN apt-get upgrade and dist-upgrade, as many of the “essential” packages from the parent images cannot upgrade inside an unprivileged container. `